openlibrary/core/booknotes.py

```diff
@@ -146,3 +146,110 @@ def remove(cls, username, work_id, edition_id=NULL_EDITION_VALUE):
+
+          notes:         A list of dictionaries containing notes for each edition.
+                         Edition IDs can be accessed by using the key "edition_id"; and
+                         notes by the key "notes".  Edition IDs are integers.  An
+                         edition ID of -1 denotes a work-level note.
+
+        Note: Once Postgres is upgraded, we can simplify the query by using
+              json_object_agg in place of json_agg:
+
+              SELECT
+                work_id,
+                json_object_agg(observation_type_id, observation_values) as observations
+              ...
+              SELECT
+                work_id,
+                json_object_agg(edition_id, notes) as notes
+
+              This will also simplify the observations and notes dictionaries,
+              reducing the number of key-value pairs to one for each dictionary.
+
+        return: A list of a patron's booknotes and observations.
+        """
+        oldb = db.get_db()
+        data = {
+            'username': username,
+            'limit': limit,
+            'offset': limit * (page - 1),
+        }
+
+        query = """
+        SELECT
+            COALESCE(obs.work_id, bn.work_id) as work_id,
+            obs.observations,
+            bn.notes
+        FROM (
+            SELECT
+                work_id,
+                json_agg(row_to_json(
+                    (SELECT r FROM
+                        (SELECT observation_type, observation_values) r)
+                    )
+                ) as observations
+            FROM (
+                SELECT
+                work_id,
+                observation_type,
+                json_agg(observation_value) as observation_values
+                FROM observations
+                WHERE username = $username
+                GROUP BY work_id, observation_type
+            ) s
+            GROUP BY work_id) obs
+        FULL OUTER JOIN (
+            SELECT
+                work_id,
+                json_agg(row_to_json(
+                    (SELECT r FROM
+                        (SELECT edition_id, notes) r)
+                    )
+                ) as notes
+            FROM booknotes
+            WHERE username = $username
+            GROUP BY work_id
+        ) bn
+        ON obs.work_id = bn.work_id
+        LIMIT $limit OFFSET $offset
+        """
+
+        return list(oldb.query(query, vars=data))
+
+    @classmethod
+    def count_patron_booknotes_and_observations(cls, username):
+        """
+        Returns a count of distinct work IDs from both the booknotes and observations
+        tables for the given username.  If a work ID is present in both tables, it is
+        only counted once.
+
+        return: A count of distinct work IDs for the given username
+        """
+        oldb = db.get_db()
+        data = {
+            'username': username
+        }
+        query = """
+            SELECT COUNT(DISTINCT(t.work_id)) FROM (
```

Comment:
I'm not sure if `DISTINCT` is needed here.  The `UNION` should only be selecting distinct values.